File: core/src/git.rs
use crate::errors::{CoreError, CoreResult};
use crate::security::WorkspaceGuard;
use regex::Regex;
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::HashSet;
use std::path::Path;
use std::process::Command;

/// CodeLink never forwards a client-supplied command string to a shell.
/// Every Git operation below builds a fixed, known argv shape; the only
/// client-controlled inputs are path and revision *values*, which are
/// validated and passed as discrete arguments (never concatenated into a
/// shell string), so there is no command or argument injection surface.
fn run_git(root: &Path, args: &[&str]) -> CoreResult<(String, String, i32)> {
    let output = Command::new("git")
        .args(args)
        .current_dir(root)
        .output()
        .map_err(|e| CoreError::new("GIT_NOT_AVAILABLE", format!("Failed to run git: {e}")))?;
    Ok((
        String::from_utf8_lossy(&output.stdout).to_string(),
        String::from_utf8_lossy(&output.stderr).to_string(),
        output.status.code().unwrap_or(-1),
    ))
}

fn ensure_success(stdout: String, stderr: String, code: i32) -> CoreResult<String> {
    if code != 0 {
        let message = if stderr.trim().is_empty() {
            stdout
        } else {
            stderr
        };
        return Err(CoreError::git_command_failed(message));
    }
    Ok(stdout)
}

fn validate_revision(rev: &str) -> CoreResult<()> {
    if rev.trim().is_empty() {
        return Err(CoreError::invalid_argument("rev must not be empty"));
    }
    if rev.starts_with('-') {
        return Err(CoreError::invalid_argument(
            "Revision must not start with '-'",
        ));
    }
    Ok(())
}

#[derive(Debug, Deserialize)]
struct RootParams {
    root: String,
}

pub fn status(params: Value) -> CoreResult<Value> {
    let p: RootParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = WorkspaceGuard::new(&p.root)?;
    let (out, err, code) = run_git(guard.root(), &["status", "--porcelain=v1", "--branch"])?;
    let out = ensure_success(out, err, code)?;

    let ahead_re = Regex::new(r"ahead (\d+)").unwrap();
    let behind_re = Regex::new(r"behind (\d+)").unwrap();

    let mut branch: Option<String> = None;
    let mut upstream: Option<String> = None;
    let mut ahead: i64 = 0;
    let mut behind: i64 = 0;
    let mut files = Vec::new();

    for line in out.lines() {
        if let Some(rest) = line.strip_prefix("## ") {
            let branch_part = rest.split(" [").next().unwrap_or(rest);
            if let Some((local, remote)) = branch_part.split_once("...") {
                branch = Some(local.to_string());
                upstream = Some(remote.to_string());
            } else {
                branch = Some(branch_part.to_string());
            }
            if let Some(caps) = ahead_re.captures(rest) {
                ahead = caps[1].parse().unwrap_or(0);
            }
            if let Some(caps) = behind_re.captures(rest) {
                behind = caps[1].parse().unwrap_or(0);
            }
        } else if line.len() >= 3 {
            files.push(json!({
                "path": line[3..],
                "indexStatus": &line[0..1],
                "worktreeStatus": &line[1..2],
            }));
        }
    }

    Ok(json!({
        "branch": branch,
        "upstream": upstream,
        "ahead": ahead,
        "behind": behind,
        "files": files,
    }))
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct DiffParams {
    root: String,
    #[serde(default)]
    path: Option<String>,
    #[serde(default)]
    staged: bool,
}

pub fn diff(params: Value) -> CoreResult<Value> {
    let p: DiffParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = WorkspaceGuard::new(&p.root)?;

    let mut args: Vec<String> = vec!["diff".to_string()];
    if p.staged {
        args.push("--staged".to_string());
    }
    if let Some(path) = &p.path {
        let resolved = guard.resolve(path)?;
        args.push("--".to_string());
        args.push(guard.to_relative_display(&resolved));
    }

    let args_ref: Vec<&str> = args.iter().map(String::as_str).collect();
    let (out, err, code) = run_git(guard.root(), &args_ref)?;
    let out = ensure_success(out, err, code)?;
    Ok(json!({ "diff": out }))
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct LogParams {
    root: String,
    #[serde(default)]
    max_count: Option<u32>,
    #[serde(default)]
    path: Option<String>,
}

const RECORD_SEP: &str = "\u{1e}";
const FIELD_SEP: &str = "\u{1f}";

pub fn log(params: Value) -> CoreResult<Value> {
    let p: LogParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = WorkspaceGuard::new(&p.root)?;
    let max_count = p.max_count.unwrap_or(20).clamp(1, 200);

    let mut args: Vec<String> = vec![
        "log".to_string(),
        format!("-{max_count}"),
        format!(
            "--pretty=format:%H{FIELD_SEP}%an{FIELD_SEP}%ae{FIELD_SEP}%aI{FIELD_SEP}%s{RECORD_SEP}"
        ),
    ];
    if let Some(path) = &p.path {
        let resolved = guard.resolve(path)?;
        args.push("--".to_string());
        args.push(guard.to_relative_display(&resolved));
    }

    let args_ref: Vec<&str> = args.iter().map(String::as_str).collect();
    let (out, err, code) = run_git(guard.root(), &args_ref)?;
    let out = ensure_success(out, err, code)?;

    let commits: Vec<Value> = out
        .split(RECORD_SEP)
        .map(str::trim)
        .filter(|record| !record.is_empty())
        .map(|record| {
            let parts: Vec<&str> = record.trim_start_matches('\n').split(FIELD_SEP).collect();
            json!({
                "hash": parts.first().copied().unwrap_or(""),
                "author": parts.get(1).copied().unwrap_or(""),
                "email": parts.get(2).copied().unwrap_or(""),
                "date": parts.get(3).copied().unwrap_or(""),
                "message": parts.get(4).copied().unwrap_or(""),
            })
        })
        .collect();

    Ok(json!({ "commits": commits }))
}

pub fn branches(params: Value) -> CoreResult<Value> {
    let p: RootParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = WorkspaceGuard::new(&p.root)?;
    let (out, err, code) = run_git(
        guard.root(),
        &[
            "branch",
            "-a",
            "--format=%(refname:short)\t%(HEAD)\t%(upstream:short)",
        ],
    )?;
    let out = ensure_success(out, err, code)?;

    let branches: Vec<Value> = out
        .lines()
        .filter(|l| !l.trim().is_empty())
        .map(|line| {
            let parts: Vec<&str> = line.split('\t').collect();
            let upstream = parts.get(2).copied().unwrap_or("");
            json!({
                "name": parts.first().copied().unwrap_or(""),
                "current": parts.get(1).copied() == Some("*"),
                "upstream": if upstream.is_empty() { Value::Null } else { Value::String(upstream.to_string()) },
            })
        })
        .collect();

    Ok(json!({ "branches": branches }))
}

#[derive(Debug, Deserialize)]
struct ShowParams {
    root: String,
    rev: String,
    #[serde(default)]
    path: Option<String>,
}

pub fn show(params: Value) -> CoreResult<Value> {
    let p: ShowParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    validate_revision(&p.rev)?;
    let guard = WorkspaceGuard::new(&p.root)?;

    let target = match &p.path {
        Some(path) => {
            let resolved = guard.resolve(path)?;
            format!("{}:{}", p.rev, guard.to_relative_display(&resolved))
        }
        None => p.rev.clone(),
    };

    let (out, err, code) = run_git(guard.root(), &["show", "--no-color", &target])?;
    let out = ensure_success(out, err, code)?;
    Ok(json!({ "content": out }))
}

pub fn remote(params: Value) -> CoreResult<Value> {
    let p: RootParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = WorkspaceGuard::new(&p.root)?;
    let (out, err, code) = run_git(guard.root(), &["remote", "-v"])?;
    let out = ensure_success(out, err, code)?;

    let mut seen = HashSet::new();
    let mut remotes = Vec::new();
    for line in out.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 && seen.insert(parts[0].to_string()) {
            remotes.push(json!({ "name": parts[0], "url": parts[1] }));
        }
    }
    Ok(json!({ "remotes": remotes }))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::process::Command as StdCommand;

    fn init_repo() -> tempfile::TempDir {
        let dir = tempfile::tempdir().unwrap();
        let run = |args: &[&str]| {
            StdCommand::new("git")
                .args(args)
                .current_dir(dir.path())
                .output()
                .expect("git available in test environment")
        };
        run(&["init", "-q"]);
        run(&["config", "user.email", "test@example.com"]);
        run(&["config", "user.name", "Test"]);
        std::fs::write(dir.path().join("a.txt"), "hello").unwrap();
        run(&["add", "."]);
        run(&["commit", "-q", "-m", "initial"]);
        dir
    }

    #[test]
    fn status_reports_clean_tree() {
        let dir = init_repo();
        let result = status(json!({ "root": dir.path().to_string_lossy() })).unwrap();
        assert_eq!(result["files"].as_array().unwrap().len(), 0);
    }

    #[test]
    fn status_reports_untracked_file() {
        let dir = init_repo();
        std::fs::write(dir.path().join("b.txt"), "new").unwrap();
        let result = status(json!({ "root": dir.path().to_string_lossy() })).unwrap();
        assert_eq!(result["files"].as_array().unwrap().len(), 1);
    }

    #[test]
    fn log_returns_initial_commit() {
        let dir = init_repo();
        let result = log(json!({ "root": dir.path().to_string_lossy() })).unwrap();
        let commits = result["commits"].as_array().unwrap();
        assert_eq!(commits.len(), 1);
        assert_eq!(commits[0]["message"], "initial");
    }

    #[test]
    fn show_rejects_flag_like_revision() {
        let dir = init_repo();
        let err = show(json!({ "root": dir.path().to_string_lossy(), "rev": "--upload-pack=x" }))
            .unwrap_err();
        assert_eq!(err.code, "INVALID_ARGUMENT");
    }

    #[test]
    fn diff_rejects_path_outside_workspace() {
        let dir = init_repo();
        let err = diff(json!({ "root": dir.path().to_string_lossy(), "path": "../outside.txt" }))
            .unwrap_err();
        assert_eq!(err.code, "PATH_OUTSIDE_WORKSPACE");
    }
}

File: extension/src/core/git.ts
import { execFile } from "node:child_process";
import { CodeLinkError, ErrorCodes } from "../utils/errors.js";
import { WorkspaceGuard } from "./workspaceGuard.js";

const MAX_BUFFER_BYTES = 64 * 1024 * 1024;

/** CodeLink never forwards a client-supplied command string to a shell.
 * Every Git operation below builds a fixed, known argv shape; the only
 * client-controlled inputs are path and revision *values*, which are
 * validated and passed as discrete arguments to `execFile` (never
 * concatenated into a shell string), so there is no command or argument
 * injection surface. */
function runGit(cwd: string, args: string[]): Promise<{ stdout: string; stderr: string; code: number }> {
  return new Promise((resolve, reject) => {
    execFile(
      "git",
      args,
      { cwd, maxBuffer: MAX_BUFFER_BYTES, windowsHide: true },
      (error, stdout, stderr) => {
        if (error && typeof (error as NodeJS.ErrnoException).code === "string" && (error as NodeJS.ErrnoException).code === "ENOENT") {
          reject(new CodeLinkError(ErrorCodes.GIT_NOT_AVAILABLE, `Failed to run git: ${error.message}`));
          return;
        }
        const code = typeof error?.code === "number" ? error.code : error ? -1 : 0;
        resolve({ stdout, stderr, code });
      },
    );
  });
}

function ensureSuccess({ stdout, stderr, code }: { stdout: string; stderr: string; code: number }): string {
  if (code !== 0) {
    const message = stderr.trim().length === 0 ? stdout : stderr;
    throw new CodeLinkError(ErrorCodes.GIT_COMMAND_FAILED, message);
  }
  return stdout;
}

function validateRevision(rev: string): void {
  if (rev.trim().length === 0) {
    throw new CodeLinkError(ErrorCodes.INVALID_ARGUMENT, "rev must not be empty");
  }
  if (rev.startsWith("-")) {
    throw new CodeLinkError(ErrorCodes.INVALID_ARGUMENT, "Revision must not start with '-'");
  }
}

export interface RootParams {
  root: string;
}

export async function status(
  params: RootParams,
): Promise<{ branch: string | null; upstream: string | null; ahead: number; behind: number; files: unknown[] }> {
  const guard = await WorkspaceGuard.create(params.root);
  const out = ensureSuccess(await runGit(guard.root(), ["status", "--porcelain=v1", "--branch"]));

  const aheadRe = /ahead (\d+)/;
  const behindRe = /behind (\d+)/;

  let branch: string | null = null;
  let upstream: string | null = null;
  let ahead = 0;
  let behind = 0;
  const files: { path: string; indexStatus: string; worktreeStatus: string }[] = [];

  for (const line of out.split("\n")) {
    if (line.startsWith("## ")) {
      const rest = line.slice(3);
      const branchPart = rest.split(" [")[0] ?? rest;
      if (branchPart.includes("...")) {
        const [local, remote] = branchPart.split("...");
        branch = local ?? "";
        upstream = remote ?? "";
      } else {
        branch = branchPart;
      }
      const aheadMatch = aheadRe.exec(rest);
      if (aheadMatch) {
        ahead = Number.parseInt(aheadMatch[1] ?? "0", 10) || 0;
      }
      const behindMatch = behindRe.exec(rest);
      if (behindMatch) {
        behind = Number.parseInt(behindMatch[1] ?? "0", 10) || 0;
      }
    } else if (line.length >= 3) {
      files.push({ path: line.slice(3), indexStatus: line.slice(0, 1), worktreeStatus: line.slice(1, 2) });
    }
  }

  return { branch, upstream, ahead, behind, files };
}

export interface DiffParams {
  root: string;
  path?: string;
  staged?: boolean;
}

export async function diff(params: DiffParams): Promise<{ diff: string }> {
  const guard = await WorkspaceGuard.create(params.root);
  const args = ["diff"];
  if (params.staged) {
    args.push("--staged");
  }
  if (params.path) {
    const resolved = await guard.resolve(params.path);
    args.push("--", guard.toRelativeDisplay(resolved));
  }
  return { diff: ensureSuccess(await runGit(guard.root(), args)) };
}

export interface LogParams {
  root: string;
  maxCount?: number;
  path?: string;
}

const RECORD_SEP = "";
const FIELD_SEP = "";

export async function log(
  params: LogParams,
): Promise<{ commits: { hash: string; author: string; email: string; date: string; message: string }[] }> {
  const guard = await WorkspaceGuard.create(params.root);
  const maxCount = Math.min(Math.max(params.maxCount ?? 20, 1), 200);

  const args = ["log", `-${maxCount}`, `--pretty=format:%H${FIELD_SEP}%an${FIELD_SEP}%ae${FIELD_SEP}%aI${FIELD_SEP}%s${RECORD_SEP}`];
  if (params.path) {
    const resolved = await guard.resolve(params.path);
    args.push("--", guard.toRelativeDisplay(resolved));
  }

  const out = ensureSuccess(await runGit(guard.root(), args));
  const commits = out
    .split(RECORD_SEP)
    .map((record) => record.trim())
    .filter((record) => record.length > 0)
    .map((record) => {
      const parts = record.replace(/^\n+/, "").split(FIELD_SEP);
      return {
        hash: parts[0] ?? "",
        author: parts[1] ?? "",
        email: parts[2] ?? "",
        date: parts[3] ?? "",
        message: parts[4] ?? "",
      };
    });

  return { commits };
}

export async function branches(
  params: RootParams,
): Promise<{ branches: { name: string; current: boolean; upstream: string | null }[] }> {
  const guard = await WorkspaceGuard.create(params.root);
  const out = ensureSuccess(
    await runGit(guard.root(), ["branch", "-a", "--format=%(refname:short)\t%(HEAD)\t%(upstream:short)"]),
  );

  const list = out
    .split("\n")
    .filter((line) => line.trim().length > 0)
    .map((line) => {
      const parts = line.split("\t");
      const upstream = parts[2] ?? "";
      return { name: parts[0] ?? "", current: parts[1] === "*", upstream: upstream.length === 0 ? null : upstream };
    });

  return { branches: list };
}

export interface ShowParams {
  root: string;
  rev: string;
  path?: string;
}

export async function show(params: ShowParams): Promise<{ content: string }> {
  validateRevision(params.rev);
  const guard = await WorkspaceGuard.create(params.root);

  let target = params.rev;
  if (params.path) {
    const resolved = await guard.resolve(params.path);
    target = `${params.rev}:${guard.toRelativeDisplay(resolved)}`;
  }

  return { content: ensureSuccess(await runGit(guard.root(), ["show", "--no-color", target])) };
}

export async function remote(params: RootParams): Promise<{ remotes: { name: string; url: string }[] }> {
  const guard = await WorkspaceGuard.create(params.root);
  const out = ensureSuccess(await runGit(guard.root(), ["remote", "-v"]));

  const seen = new Set<string>();
  const remotes: { name: string; url: string }[] = [];
  for (const line of out.split("\n")) {
    const parts = line.split(/\s+/).filter(Boolean);
    const name = parts[0];
    const url = parts[1];
    if (name !== undefined && url !== undefined && !seen.has(name)) {
      seen.add(name);
      remotes.push({ name, url });
    }
  }
  return { remotes };
}
